working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+   _as is_ within isolates. Newly written code can choose to opt-in into
+   shareability where it matters for performance or interoperability reasons.
+   Marking class as `Shareable` gives a strong signal that the developer has
+   considered implications of sharing instances of these class across threads
+   and took measure to ensure that it is safe.
+
+I would like to introduce shared memory multithreading into Dart in a way that
+avoids subtle breakages in the existing code. Consider for a moment a library
+which maintains a global cache internally and is written under the assumption
+that Dart is a single threaded:
+
+```dart
+class C {
+  final id;
+  C(this.id);
+}
+
+int _nextId = 0;
+final _cache = <int, C>{};
+
+C makeObject() {
+  _cache[nextId] = C(nextId);
+  nextId++;
+  return _cache[nextId - 1];
+}
+```
+
+This was a valid way to structure this code in single-threaded Dart, but the
+same code becomes thread _unsafe_ in the presence of shared memory
+multithreading.
+
+### Generics
+
+When declaring a generic type the developer will have to use type parameter
+bounds to ensure that resulting class conforms to restrictions imposed by
+`Shareable`:
+
+```dart
+class X<T> implements Shareable {
+  T v;  // compile time error.
+}
+
+class Y<T extends Shareable> implements Shareable {
+  T v;  // ok.
+}
+```
+
+> [!NOTE]
+>
+> We could really benefit from the ability to use intersection types here, which
+> would allow to specify complicated bounds like `T extends Shareable & I`. In
+> the absence of intersections types developers would be forced to declare
+> intermediate interfaces which implement all required interfaces (e.g.
+> `abstract interface ShareableI implements Shareable, I {}`) and require users
+> to implement those by specifying `T extends ShareableI`.
+
+### Functions
+
+Shareability of a function depends on the values that it captures. We could
+define that any **function is shareable iff it captures only variables of
+shareable type**. Incorporating this property into the type system naturally
+leads to the desire to use _intersection types_ to express the property that
+some value is both a function of a specific type _and_ shareable:
+
+```dart
+class A implements Shareable {
+  void Function() f;  // compile time error
+  void Function() & Shareable f;  // ok
+}
+```
+
+Introducing intersection types into type system might be a huge undertaking. For
+the purposes of developing an MVP we can choose one of the two approaches:
+
+- Ignore functions entirely: consider functions un-shareable. It becomes a
+  compile time error to have a function type field inside a shareable class.
+- Allow function type fields inside a shareable class, but enforce shareability
+  in runtime on assignment to the field.
+
+### Shareable core types
+
+The following types will become shareable:
+
+- `num` (`int` and `double`), `String`, `bool`, `Null`, `BigInt`
+- `Enum` and consequently all user defined enums
```

Comment:
Good point. 

I wonder if we should consider a variation of _construction time immutability_ rule, e.g. 

```
/// Mark the class as deeply immutable after initialization. 
/// 
/// If class is marked as deeply immutable then:
///
///   * all of its fields must be non-late final
///   * fields should be initialized with values which are assignable to `DeeplyImmutable`
///  
/// All constant values are implicitly subtypes of `DeeplyImmutable` even if
/// their class is not marked as such.
abstract interface class DeeplyImmutable implements Shareable {
}
```

It's a bit awkward, but allows to say that `Enum implements DeeplyImmutable` and create other values which are deeply immutable.

WDYT?

Reply:
If it's a type, then we have some instances of a class (constants) implementing an interface while other instances do not. That's effectively as if constants were instances of a secret subclass which implements , not the real class.

Which could be the story, if a `class` has a `const` constructor, there is an implicit synthetic `class _$TheClass<T> extends TheClass<T> with DeeplyImmutable {Type get runtimeType => TheClass<T>;}`, which constant constructor invocations are creating instances of. (But not `enums` which implement `DeeplyImmutable` directly.)

The type system would have to know that, and know that the type of any constant expression with static type `Foo` is really `Foo & DeeplyImmutable`, which will be a thing. (Can't assume that _any_ type has a deeply immutable subtype.)

It does mean that we can now distinguish constant values from non-constant values with the same state. At least if the class isn't deeply immutable anyway.

That's cumbersome, and any class which declares itself as `implements DeeplyImmutable` forces that on all subclasses, and like `Shareable`, it requires all supertypes to be, and stay, compatible.
Having a const constructor is not enough to allow a subclass to declare itself `DeeplyImmutable` and have a non-constant constructor.

Maybe it's better to introduce `const Foo` as a type modifier, which introduces a subtype, but not a subclass. Then any `const` type is sendable, but not `Sendable`, just like `null`.
Does preclude guarding by type, since the demand objects are then subtypes of either `Sendable` or `const Object?`, but those have no immediate union type.

All in all, not trivial.

Reply:
If the only goal of `DeeplyImmutable` is to be sendable, it's probably easier to just say that constant values implement `Sendable`, with some of the same strategies.
So `const Foo()` would have type `const Foo`, which is assignable to `Foo` and to `Sendable`. It's the intersection type, basically

It is still a type where some values are `Sendable` and others are not. We don't have intersection types, so one cannot ask for `Foo & Sendable` and accept any sendable subtype of `Foo` (might have done subtypes which are sendable and others which are not). This approach only allows to single out the `const` values.

Reply:
I think `DeeplyImmutable` as a concept is useful.

1. `Isolate.send` will copy everything mutable but share `DeeplyImmutable` objects.
2. `Isolate.share` can only share object graphs that are transitively `Shareable` or `DeeplyImmutable`. (Maybe `DeeplyImmutable implements Sharable`.)

Both concepts are useful for `NativeFinalizer`s, they both allow running a native finalizer only after all isolates in an isolate group all lose access to a `Finalizable`.

`Sendable` as a concept does not enable `Isolate.send`ing deeply immutable `Finalizable`s. So, I believe it would be useful to introduce `DeeplyImmutable` as a concept. (Whether it should be a `Type` or a vm-pragma is another discussion.)

Context: https://github.com/dart-lang/sdk/issues/54885#issuecomment-1966427560